### PR TITLE
Chore in discrete adjoint bits

### PR DESCRIPTION
**Pull request by TobiKattmann**
Hi all,

As I am stepping through DA cases with my debugger I am cleaning some things left and right. No major changes.

- making local vars const when possible
- moving the `config->GetBool` directly to the conditional instead of making it a var and using it 1 time
- removing dead code
- adding some comments
- some typos fixed

## PR Checklist

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags, or simply --warnlevel=2 when using meson).
- [x] My contribution is commented and consistent with SU2 style.
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

### CODE REVIEW

File: SU2_CFD/src/solvers/CDiscAdjSolver.cpp [hunk 1]
```diff
@@ -376,7 +371,7 @@ void CDiscAdjSolver::ExtractAdjoint_Solution(CGeometry *geometry, CConfig *confi
 
   if (time_n_needed) {
     SU2_OMP_FOR_STAT(omp_chunk_size)
-    for (auto iPoint = 0u; iPoint < nPoint; iPoint++) {
+    for (auto iPoint = 0ul; iPoint < nPoint; iPoint++) {
```
Comment (TobiKattmann) [inline SU2_CFD/src/solvers/CDiscAdjSolver.cpp:374]: this  iPoint change from 0u to 0ul is less of a necessity then rather a consistency thing

File: SU2_CFD/include/variables/CDiscAdjVariable.hpp [hunk 2]
```diff
@@ -75,40 +76,69 @@ class CDiscAdjVariable final : public CVariable {
...
   inline void SetDual_Time_Derivative_n(unsigned long iPoint, unsigned long iVar, su2double der) override {
     DualTime_Derivative_n(iPoint,iVar) = der;
   }
 
-  // TK:: Description missing
+  /*!
+   * \brief Return the dual time contributions to the adjoint variable. Contains contributions from 2 timesteps for dual time 2nd order.
+   * \param[in] iPoint - Point index
+   * \param[in] iVar - Index of the variable
+   * \return Derivative value
```
Comment (pcarruscag) [inline SU2_CFD/include/variables/CDiscAdjVariable.hpp:111]: Kudos for the briefs.
The params and returns on the other hand... I think they are superfluous, they just make you scroll more...
Moreover they are repeated a bunch of times, hence should be avoided.
"The only documentation that stays up to date is the one the compiler understands"

File: SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp [hunk 3]
```diff
@@ -421,6 +421,8 @@ void CDiscAdjMultizoneDriver::Run() {
         }
       }
 
+      // TobiKattmann:: Add dual time terms dG/du from n+1 and N+2 to external
```
Comment (pcarruscag) [inline SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:424]: I think there is an open PR for that kind of stuff :)

File: SU2_CFD/include/variables/CDiscAdjVariable.hpp (hunk 2 above)
Comment (TobiKattmann) [reply SU2_CFD/include/variables/CDiscAdjVariable.hpp:111]: I guess you're right. I'll remove the superfluous stuff

File: SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp (hunk 3 above)
Comment (TobiKattmann) [reply SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:424]: Yes, and i sure need to take a more in depth look at the PR. This is leftover of my own try to think that trough and will be removed

Comment (pcarruscag) [reply SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:424]: I think your thinking of it is very good, and aligned with Ole's idea with the external that combines the OF gradient with MZ cross terms.
All that would be needed is for `CDiscAdjSolver::SetAdjoint_Output` to not add the dual time contribution while setting the adjoints.

Comment (TobiKattmann) [reply SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:424]: Yes, but on the other hand one could keep the same logic as for single zone to some extent, which  would be nice

Comment (pcarruscag) [reply SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:424]: That would not work with GMRES applied to the adjoint, and that is something we want :)

File: SU2_CFD/src/drivers/CMultizoneDriver.cpp [hunk 4]
```diff
@@ -692,11 +663,9 @@ bool CMultizoneDriver::Monitor(unsigned long TimeIter){
       cout << "-------------------------------------------------------------------------" << endl;
     }
 
-    StopCalc = FinalTimeReached || MaxIterationsReached;
+    return (FinalTimeReached || MaxIterationsReached);
   }
 
-  return StopCalc;
```
Comment (TobiKattmann) [inline SU2_CFD/src/drivers/CMultizoneDriver.cpp:698]: StopCalc is class data and was returned to itself, so I made the function to not use that var

File: SU2_CFD/src/drivers/CMultizoneDriver.cpp [hunk 5]
```diff
@@ -237,25 +230,15 @@ void CMultizoneDriver::Preprocess(unsigned long TimeIter) {
...
       config_container[iZone]->SetPhysicalTime(static_cast<su2double>(TimeIter)*config_container[iZone]->GetDelta_UnstTimeND());
-    else
+    }
+    else {
       config_container[iZone]->SetPhysicalTime(0.0);
-
-    /*--- Read the target pressure for inverse design. ---------------------------------------------*/
-    /*--- TODO: This routine should be taken out of output, and made general for multiple zones. ---*/
-//    if (config_container[iZone]->GetInvDesign_Cp() == YES)
-//      output[iZone]->SetCp_InverseDesign(solver_container[iZone][INST_0][MESH_0][FLOW_SOL],
```
Comment (TobiKattmann) [inline SU2_CFD/src/drivers/CMultizoneDriver.cpp:248]: I removed all  the inverse design stuff that was commented out in multiple places

File: SU2_CFD/src/iteration/CDiscAdjHeatIteration.cpp [hunk 6]
```diff
@@ -32,25 +32,21 @@ void CDiscAdjHeatIteration::Preprocess(COutput* output, CIntegration**** integra
...
+  const bool dual_time_2nd = (config[val_iZone]->GetTime_Marching() == TIME_MARCHING::DT_STEPPING_2ND);
+  const bool dual_time = (dual_time_1st || dual_time_2nd);
+
+  auto solvers = solver[val_iZone][val_iInst];
 
   /*--- For the unsteady adjoint, load direct solutions from restart files. ---*/
 
   if (config[val_iZone]->GetTime_Marching() != TIME_MARCHING::STEADY) {
-    Direct_Iter = SU2_TYPE::Int(config[val_iZone]->GetUnst_AdjointIter()) - SU2_TYPE::Int(TimeIter) - 2;
+    const int Direct_Iter = SU2_TYPE::Int(config[val_iZone]->GetUnst_AdjointIter()) - SU2_TYPE::Int(TimeIter) - 2 + dual_time;
```
Comment (TobiKattmann) [inline SU2_CFD/src/iteration/CDiscAdjHeatIteration.cpp:46]: now conistent with CDiscAdjFluidIteration.cpp

**Comment by TobiKattmann** [comment]
@cvencro Merging this will most likely result in some conflicts with #1260 which are hopefully easy to resolve as it is mostly stylistic stuff without any real changes to the code. @pcarruscag I reverted CDiscADjVariable.hpp/cpp and will readd the briefs in #1260

File: SU2_CFD/src/drivers/CDriver.cpp [hunk 7]
```diff
@@ -288,10 +287,10 @@ CDriver::CDriver(char* confFile, unsigned short val_nZone, SU2_Comm MPICommunica
   Mpoints            = 0.0;
   MpointsDomain      = 0.0;
   for (iZone = 0; iZone < nZone; iZone++) {
-    Mpoints       +=(su2double)geometry_container[iZone][INST_0][MESH_0]->GetGlobal_nPoint()/(1.0e6);
-    MpointsDomain +=(su2double)geometry_container[iZone][INST_0][MESH_0]->GetGlobal_nPointDomain()/(1.0e6);
-    MDOFs         += (su2double)DOFsPerPoint*(su2double)geometry_container[iZone][INST_0][MESH_0]->GetGlobal_nPoint()/(1.0e6);
-    MDOFsDomain   += (su2double)DOFsPerPoint*(su2double)geometry_container[iZone][INST_0][MESH_0]->GetGlobal_nPointDomain()/(1.0e6);
+    Mpoints       += static_cast<su2double>(geometry_container[iZone][INST_0][MESH_0]->GetGlobal_nPoint())/(1.0e6);
+    MpointsDomain += static_cast<su2double>(geometry_container[iZone][INST_0][MESH_0]->GetGlobal_nPointDomain())/(1.0e6);
+    MDOFs         += static_cast<su2double>(DOFsPerPoint)*static_cast<su2double>(geometry_container[iZone][INST_0][MESH_0]->GetGlobal_nPoint())/(1.0e6);
```
Comment (pcarruscag) [inline SU2_CFD/src/drivers/CDriver.cpp:292]: The casts are not really needed.

File: SU2_CFD/src/iteration/CDiscAdjHeatIteration.cpp (hunk 6 above)
Comment (pcarruscag) [reply SU2_CFD/src/iteration/CDiscAdjHeatIteration.cpp:46]: Fun fact, SU2_TYPE::Int implies a cast to su2double, followed cast to int, yeyy

File: SU2_CFD/src/solvers/CDiscAdjSolver.cpp [hunk 8]
```diff
@@ -477,25 +466,31 @@ void CDiscAdjSolver::SetAdjoint_Output(CGeometry *geometry, CConfig *config) {
...
     }
+
+    /*--- Add dual time contributions to the adjoint solution. Two terms stored for DT-2nd-order. ---*/
     if (dual_time) {
       for (auto iVar = 0u; iVar < nVar; iVar++) {
         Solution[iVar] += nodes->GetDual_Time_Derivative(iPoint,iVar);
       }
     }
+
+    /*--- Set the adjoint values of the primal (TK::??) solution. ---*/
```
Comment (pcarruscag) [inline SU2_CFD/src/solvers/CDiscAdjSolver.cpp:486]: yes

File: SU2_CFD/src/solvers/CHeatSolver.cpp [hunk 9]
```diff
@@ -845,7 +845,7 @@ void CHeatSolver::BC_Inlet(CGeometry *geometry, CSolver **solver_container,
           dist_ij += (Coord_j[iDim]-Coord_i[iDim])*(Coord_j[iDim]-Coord_i[iDim]);
         dist_ij = sqrt(dist_ij);
 
-        dTdn = -(nodes->GetSolution(Point_Normal,0) - Twall)/dist_ij;
+        dTdn = -(nodes->GetTemperature(Point_Normal) - Twall)/dist_ij;
```
Comment (pcarruscag) [inline SU2_CFD/src/solvers/CHeatSolver.cpp:848]: :+1:

File: SU2_CFD/src/iteration/CDiscAdjHeatIteration.cpp (hunk 6 above)
Comment (TobiKattmann) [reply SU2_CFD/src/iteration/CDiscAdjHeatIteration.cpp:46]: what are those casts via SU2_TYPE methods even for

File: SU2_CFD/src/solvers/CDiscAdjSolver.cpp (hunk 8 above)
Comment (TobiKattmann) [reply SU2_CFD/src/solvers/CDiscAdjSolver.cpp:486]: yes

File: SU2_CFD/src/iteration/CDiscAdjHeatIteration.cpp [hunk 10]
```diff
@@ -43,7 +43,7 @@ void CDiscAdjHeatIteration::Preprocess(COutput* output, CIntegration**** integra
   /*--- For the unsteady adjoint, load direct solutions from restart files. ---*/
 
   if (config[val_iZone]->GetTime_Marching() != TIME_MARCHING::STEADY) {
-    const int Direct_Iter = SU2_TYPE::Int(config[val_iZone]->GetUnst_AdjointIter()) - SU2_TYPE::Int(TimeIter) - 2 + dual_time;
+    const int Direct_Iter = static_cast<int>(config[val_iZone]->GetUnst_AdjointIter()) - static_cast<int>(TimeIter) - 2 + dual_time;
```
Comment (pcarruscag) [inline SU2_CFD/src/iteration/CDiscAdjHeatIteration.cpp]: :+1:

**Review by pcarruscag — APPROVED** [review]

**Pull request merged**